### PR TITLE
ENH: Add ZT fastpath for addmm

### LINKED ISSUE #72345: Add ZeroTensor fastpath for torch.addmm
### 🚀 The feature, motivation and pitch

see https://github.com/pytorch/pytorch/issues/69687

example PR: https://github.com/pytorch/pytorch/pull/71129

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by HaoZeke**
Fixes #72345 eventually.
- [x] CPU Implementation
- [x] ~Device Implementations~ Dispatch key
- [ ] Benchmark
- [ ] Test?
- [x] Solicit reviews
  - [x] Internal
  - [x] External

### CODE REVIEW

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 1]
```diff
@@ -1221,6 +1222,33 @@ static void addmm_impl_cpu_(
   }
 }
 
+// TODO: This should work through the structured kernel approach of TORCH_IMPL_FUNC
+// TODO: ensure devices are handled
+Tensor addmm_zerotensor(const Tensor& self, const Tensor& mat1, const Tensor& mat2, const Scalar& beta, const Scalar& alpha){
+  TORCH_INTERNAL_ASSERT(self._is_zerotensor() || mat1._is_zerotensor() || mat2._is_zerotensor() || beta.equal(0) || alpha.equal(0));
```
Comment (HaoZeke) [inline aten/src/ATen/native/LinearAlgebra.cpp:1228]: Is this something we shouldn't need?

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 2]
```diff
@@ -1221,6 +1222,33 @@ static void addmm_impl_cpu_(
...
+    result.resize_as_(self);
+    result.copy_(self);
+    return beta.equal(1) ? result : result.mul_(beta);
+  }
+  // alpha*(mat1@mat2)
+  else if (self._is_zerotensor() || beta.equal(0)){
+    result = at::matmul(mat1, mat2);
+    return alpha.equal(1) ? result : result.mul_(alpha);
+  } else /* can't reach, see assert */{
+    return at::_efficientzerotensor(self_sizes, self.options());
```
Comment (HaoZeke) [inline aten/src/ATen/native/LinearAlgebra.cpp:1248]: Without the `assert`, we will need to dispatch according to the appropriate device right?

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 1 above)
Comment (anjali411) [reply aten/src/ATen/native/LinearAlgebra.cpp:1228]: this is not needed since `addmm_zerotensor` would only be called when one of the inputs is a ZT

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 3]
```diff
@@ -1221,6 +1222,33 @@ static void addmm_impl_cpu_(
   }
 }
 
+// TODO: This should work through the structured kernel approach of TORCH_IMPL_FUNC
+// TODO: ensure devices are handled
+Tensor addmm_zerotensor(const Tensor& self, const Tensor& mat1, const Tensor& mat2, const Scalar& beta, const Scalar& alpha){
+  TORCH_INTERNAL_ASSERT(self._is_zerotensor() || mat1._is_zerotensor() || mat2._is_zerotensor() || beta.equal(0) || alpha.equal(0));
+  const auto self_sizes = self.sizes();
+  Tensor result;
+  // beta*self + alpha*(mat1@mat2) = 0
```
Comment (anjali411) [inline aten/src/ATen/native/LinearAlgebra.cpp:1231]: we should perform the size and dim checks before this

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 4]
```diff
@@ -1221,6 +1222,33 @@ static void addmm_impl_cpu_(
...
+  }
+  // beta*self
+  else if (mat1._is_zerotensor() || mat2._is_zerotensor() || alpha.equal(0)){
+    result.resize_as_(self);
+    result.copy_(self);
+    return beta.equal(1) ? result : result.mul_(beta);
+  }
+  // alpha*(mat1@mat2)
+  else if (self._is_zerotensor() || beta.equal(0)){
+    result = at::matmul(mat1, mat2);
```
Comment (anjali411) [inline aten/src/ATen/native/LinearAlgebra.cpp:1245]:
```suggestion
    result = at::mm(mat1, mat2);
```

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 2 above)
Comment (anjali411) [reply aten/src/ATen/native/LinearAlgebra.cpp:1248]: we can never reach here so we don't need this condition

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 1 above)
Comment (HaoZeke) [reply aten/src/ATen/native/LinearAlgebra.cpp:1228]: Ideally we'd like to call into this if the scalars are zero as well.

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 3 above)
Comment (anjali411) [reply aten/src/ATen/native/LinearAlgebra.cpp:1231]: https://github.com/pytorch/pytorch/blob/5375b2e994ee666996a47526c7f9629e85acfc54/aten/src/ATen/native/LinearAlgebra.cpp#L1107-L1119

Comment (anjali411) [reply aten/src/ATen/native/LinearAlgebra.cpp:1231]: I'd suggest creating a helper function that can be called from the cpu kernel as well as ZT kernel

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 5]
```diff
@@ -1221,6 +1235,31 @@ static void addmm_impl_cpu_(
...
+  else if (mat1._is_zerotensor() || mat2._is_zerotensor() || alpha.equal(0)){
+    result.resize_as_(self);
+    result.copy_(self);
+    return beta.equal(1) ? result : result.mul_(beta);
+  }
+  // alpha*(mat1@mat2)
+  else if (self._is_zerotensor() || beta.equal(0)){
+    result = at::mm(mat1, mat2);
+    return alpha.equal(1) ? result : result.mul_(alpha);
+  } else /* never reached */{
```
Comment (anjali411) [inline aten/src/ATen/native/LinearAlgebra.cpp:1258]: else is not needed

**Review by anjali411 — APPROVED** [review]

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 5 above)
Comment (HaoZeke) [reply aten/src/ATen/native/LinearAlgebra.cpp:1258]: Without the `else` here the compiler will error out as a non-void function without a return.

This seemed more clear than leaving the `else` and keeping the `return`.

**Comment by anjali411** [comment]
it'd be cool to do some benchmarking!

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 5 above)
Comment (Lezcano) [reply aten/src/ATen/native/LinearAlgebra.cpp:1258]: What about making the last else if an else and leaving the condition as a comment inside of it?

**Comment by anjali411** [comment]
Here's the benchmarking script that I wrote a while ago:
```
import torch
import itertools
import time
from torch.utils.benchmark import Timer
from torch.utils.benchmark import Compare
import sys
import pickle
import torch.autograd.forward_ad as fwAD

print('Using pytorch %s' % (torch.__version__))

shapes = [(128,), (256, 128), (2, 512, 256), (2, 64, 256, 128), (4, 2, 512, 256, 128)]
results = []
repeats = 10
device = 'cpu'
dtype = torch.double

for device in ['cpu', 'cuda']:
    for mat1_shape in shapes:
        self_ = torch.rand(*mat1_shape, dtype=dtype, device=device, requires_grad=True)
        other_ = torch.randn(*mat1_shape, dtype=dtype, device=device, requires_grad=True)
        with fwAD.dual_level():
            self_dual_obj = fwAD.make_dual(self_, torch.randn_like(self_))
            other_dual_obj = other_#fwAD.make_dual(other_, torch.zeros_like(other_))

        def fn(self_, other_):
            with fwAD.dual_level():
                out=torch.div(self_, other_)

        tasks = [("fn(self_dual_obj, other_dual_obj)", "Using ZT")]
        timers = [Timer(stmt=stmt, label=f"FWD mode AD input dtype:{dtype} device:{device}", sub_label=f"{(mat1_shape)}", description=desc, globals=globals()) for stmt, desc in tasks]

        for i, timer in enumerate(timers * repeats):
            results.append(
                timer.blocked_autorange()
            )
            print(f"\r{i + 1} / {len(timers) * repeats}", end="")
            sys.stdout.flush()

with open('with_zt.pkl', 'wb') as f:
    pickle.dump(results, f)

# with open('with_zt.pkl', 'rb') as f:
#     after_results = pickle.load(f)

# with open('without_zt.pkl', 'rb') as f:
#     before_results = pickle.load(f)

# comparison = Compare(after_results + before_results)
# comparison.print()
```

**Comment by HaoZeke** [comment]
Based on CPU results [Debug] with @anjali411's script I get:
- With ZT (1.12.0a0+gitc9e6584)
- Without (1.12.0a0+git631f035)

```
[ FWD mode AD input dtype:torch.float64 device:cpu ]                                                                                                                  
                             |   Using ZT                                                                                                                             
1 threads: ------------------------------                                                                                                                             
      (128,)                 |       44.6                                                                                                                             
      (256, 128)             |      565.2                                                                                                                             
      (2, 512, 256)          |     4035.0                                                                                                                             
      (2, 64, 256, 128)      |    68232.3 
      (4, 2, 512, 256, 128)  |  2484228.1 

Times are in microseconds (us).
```

**Comment by anjali411** [comment]
> Based on Debug builds with @anjali411's script I get:
> 
> * With ZT (1.12.0a0+gitc9e6584)
> * Without (1.12.0a0+git631f035)
> 
> ```
> [-- FWD mode AD input dtype:torch.float64 device:cpu --]
>                              |   Using ZT  |  Without ZT
> 1 threads: ---------------------------------------------
>       (128,)                 |       44.1  |       43.6 
>       (256, 128)             |      566.3  |      564.8 
>       (2, 512, 256)          |     4037.8  |     4022.1 
>       (2, 64, 256, 128)      |    67910.1  |    67825.9 
>       (4, 2, 512, 256, 128)  |  2486045.9  |  2470446.2 
> 
> Times are in microseconds (us).
> ```

which device is this for?

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 5 above)
Comment (HaoZeke) [reply aten/src/ATen/native/LinearAlgebra.cpp:1258]: Yup that'd work.

Comment (anjali411) [reply aten/src/ATen/native/LinearAlgebra.cpp:1258]: I second @Lezcano

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 6]
```diff
@@ -1221,6 +1235,31 @@ static void addmm_impl_cpu_(
...
+  if ((mat1._is_zerotensor() || mat2._is_zerotensor() || alpha.equal(0)) &&
+      (self._is_zerotensor() || beta.equal(0))) {
+    result = at::_efficientzerotensor(self_sizes, self.options());
+    return result;
+  }
+  // beta*self
+  else if (mat1._is_zerotensor() || mat2._is_zerotensor() || alpha.equal(0)){
+    result.resize_as_(self);
+    result.copy_(self);
+    return beta.equal(1) ? result : result.mul_(beta);
```
Comment (anjali411) [inline aten/src/ATen/native/LinearAlgebra.cpp:1252]:
```suggestion
    return beta.equal(1) ? self.clone() : self * beta;
```

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 7]
```diff
@@ -1221,6 +1235,31 @@ static void addmm_impl_cpu_(
...
+  }
+  // alpha*(mat1@mat2)
+  else if (self._is_zerotensor() || beta.equal(0)){
+    result = at::mm(mat1, mat2);
+    return alpha.equal(1) ? result : result.mul_(alpha);
```
Comment (anjali411) [inline aten/src/ATen/native/LinearAlgebra.cpp:1255]:
```suggestion
    return alpha.equal(1) ? result : result * alpha;
```

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 8]
```diff
@@ -1221,6 +1235,31 @@ static void addmm_impl_cpu_(
   }
 }
 
+Tensor addmm_zerotensor(const Tensor& self, const Tensor& mat1, const Tensor& mat2, const Scalar& beta, const Scalar& alpha){
+  check_addmm(self, mat1, mat2);
+  const auto self_sizes = self.sizes();
+  Tensor result;
+  // beta*self + alpha*(mat1@mat2) = 0
+  if ((mat1._is_zerotensor() || mat2._is_zerotensor() || alpha.equal(0)) &&
+      (self._is_zerotensor() || beta.equal(0))) {
```
Comment (anjali411) [inline aten/src/ATen/native/LinearAlgebra.cpp:1244]: remove `beta.equal(0)`

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 9]
```diff
@@ -1221,6 +1235,30 @@ static void addmm_impl_cpu_(
...
+    result = at::_efficientzerotensor(self_sizes, self.options());
+    return result;
+  }
+  // beta*self
+  else if (mat1._is_zerotensor() || mat2._is_zerotensor() || alpha.equal(0)){
+    return beta.equal(1) ? self.clone() : self * beta;
+  }
+  // alpha*(mat1@mat2)
+  else if (self._is_zerotensor() || beta.equal(0)){
+    result = at::mm(mat1, mat2);
```
Comment (HaoZeke) [inline aten/src/ATen/native/LinearAlgebra.cpp:1254]: Offline discussion:
--> @anjali411 profiled and found this to be the bottleneck

Also `at::mm` basically calls `at::addmm` so there is no speed up..... yet.

**Comment by HaoZeke** [comment]
Offline discussion (@anjali411) thoughts:
- Worst case --> calling `at::mm`
   - **reason?** `at::mm` calls `addmm` so this is completely wasteful
   - 4x slower than regular calls
Need to do:
- [ ] CUDA benchmarking data for current changeset (dispatch key)
- [ ] Register ZT CPU, CUDA kernels for `addmm` as fallthrough (older approach)
- [ ] Profiling

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 9 above)
Comment (HaoZeke) [reply aten/src/ATen/native/LinearAlgebra.cpp:1254]: Actually the dispatcher could just be bypassed with `at::cpu::mm` perhaps, that might fix this.

**Comment by HaoZeke** [comment]
With the non-dispatch key variant:
```
50 / 50[-------------------------- FWD mode AD input dtype:torch.float64 device:cpu -------------------------]
                    |  ZT->inp, mat1, mat2  |  ZT->mat1, mat2  |   ZT->mat2  |   ZT->inp   |    No ZT  
1 threads: --------------------------------------------------------------------------------------------
      (32, 32)      |            63.8       |         64.2     |       64.4  |       64.5  |       64.3
      (64, 64)      |            96.6       |         97.9     |       97.9  |       97.7  |       97.9
      (128, 128)    |           253.4       |        255.1     |      254.8  |      255.0  |      254.8
      (256, 256)    |          1083.4       |       1084.1     |     1083.3  |     1085.8  |     1083.9
      (512, 512)    |          5200.0       |       5186.4     |     5189.7  |     5190.4  |     5191.5
      (1024, 1024)  |         30120.6       |      30130.5     |    30153.3  |    30160.7  |    30147.9
      (2048, 2048)  |        213649.9       |     213180.5     |   212828.0  |   213251.8  |   213526.8
      (4096, 4096)  |       1492927.4       |    1495480.3     |  1493053.7  |  1493534.2  |  1495705.4
```

File: aten/src/ATen/native/cuda/Blas.cpp [hunk 10]
```diff
@@ -181,6 +181,19 @@ Tensor& addmm_out_cuda_impl(Tensor& result, const Tensor& self, const Tensor& ma
   }
 
   if (&result != &self) {
+
+    // beta*self + alpha*(mat1@mat2) = 0
+    if ((mat1._is_zerotensor() || mat2._is_zerotensor() || alpha.equal(0.0)) &&
```
Comment (anjali411) [inline aten/src/ATen/native/cuda/Blas.cpp:186]: If alpha and beta are passed as zero by a user, this would surprisingly give them an efficient zero tensor. This is not a desired behavior for now since Zero Tensors are not exposed in the public API right now. I'd suggest removing the check for alpha and beta

File: aten/src/ATen/native/cuda/Blas.cpp [hunk 11]
```diff
@@ -181,6 +181,19 @@ Tensor& addmm_out_cuda_impl(Tensor& result, const Tensor& self, const Tensor& ma
...
+        (self._is_zerotensor() || beta.equal(0.0))) {
+      result = at::_efficientzerotensor(self_sizes, self.options());
+      return result;
+    }
+
+    // beta*self
+    if (mat1._is_zerotensor() || mat2._is_zerotensor() || alpha.equal(0.0)){
```
Comment (anjali411) [inline aten/src/ATen/native/cuda/Blas.cpp:193]: same comment as above for removing the check for alpha

File: aten/src/ATen/native/cuda/Blas.cpp [hunk 12]
```diff
@@ -181,6 +181,19 @@ Tensor& addmm_out_cuda_impl(Tensor& result, const Tensor& self, const Tensor& ma
   }
 
   if (&result != &self) {
+
```
Comment (anjali411) [inline aten/src/ATen/native/cuda/Blas.cpp:184]: please add a check above to verify that result is not a zero tensor

File: aten/src/ATen/native/cuda/Blas.cpp [hunk 13]
```diff
@@ -181,6 +181,19 @@ Tensor& addmm_out_cuda_impl(Tensor& result, const Tensor& self, const Tensor& ma
...
+        (self._is_zerotensor() || beta.equal(0.0))) {
+      result = at::_efficientzerotensor(self_sizes, self.options());
+      return result;
+    }
+
+    // beta*self
+    if (mat1._is_zerotensor() || mat2._is_zerotensor() || alpha.equal(0.0)){
+      return beta.equal(1) ? self.clone() : self * beta;
+    }
+
```
Comment (anjali411) [inline aten/src/ATen/native/cuda/Blas.cpp:196]: these checks should be before the `if (&result != &self) {` condition

File: aten/src/ATen/native/cuda/Blas.cpp [hunk 14]
```diff
@@ -181,6 +181,19 @@ Tensor& addmm_out_cuda_impl(Tensor& result, const Tensor& self, const Tensor& ma
...
+    // beta*self + alpha*(mat1@mat2) = 0
+    if ((mat1._is_zerotensor() || mat2._is_zerotensor() || alpha.equal(0.0)) &&
+        (self._is_zerotensor() || beta.equal(0.0))) {
+      result = at::_efficientzerotensor(self_sizes, self.options());
+      return result;
+    }
+
+    // beta*self
+    if (mat1._is_zerotensor() || mat2._is_zerotensor() || alpha.equal(0.0)){
+      return beta.equal(1) ? self.clone() : self * beta;
```
Comment (anjali411) [inline aten/src/ATen/native/cuda/Blas.cpp:194]:
```suggestion
      return self * beta;
```

since we are cloning anyway

File: aten/src/ATen/native/cuda/Blas.cpp (hunk 11 above)
Comment (anjali411) [inline aten/src/ATen/native/cuda/Blas.cpp:193]:
```suggestion
    if (mat1._is_zerotensor() || mat2._is_zerotensor() || alpha.toComplexDouble() == 0.0){
```

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 15]
```diff
@@ -1134,6 +1134,21 @@ static void addmm_impl_cpu_(
     return;
   }
 
+  // beta*self + alpha*(mat1@mat2) = 0
```
Comment (anjali411) [inline aten/src/ATen/native/LinearAlgebra.cpp:1137]: please add a check to verify that result is not a zero tensor

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 16]
```diff
@@ -1134,6 +1134,21 @@ static void addmm_impl_cpu_(
...
+  // Note that alpha*(mat1@mat2) calls into mm so is not special cased
+  if ((m1._is_zerotensor() || m2._is_zerotensor() || alpha.equal(0.0)) &&
```
Comment (anjali411) [inline aten/src/ATen/native/LinearAlgebra.cpp:1139]: same comment as below (remove the checks for alpha and beta)

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 17]
```diff
@@ -1134,6 +1134,21 @@ static void addmm_impl_cpu_(
...
+      (self._is_zerotensor() || beta.equal(0.0))) {
+    result = at::_efficientzerotensor(self_sizes, self.options());
```
Comment (anjali411) [inline aten/src/ATen/native/LinearAlgebra.cpp:1141]:
```suggestion
    return at::_efficientzerotensor(self_sizes, self.options());
```

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 18]
```diff
@@ -1134,6 +1134,21 @@ static void addmm_impl_cpu_(
...
+    return;
+  }
+
+  // beta*self
+  if (m1._is_zerotensor() || m2._is_zerotensor() || alpha.equal(0.0)){
+    result = beta.equal(1) ? self.clone() : self * beta;
```
Comment (anjali411) [inline aten/src/ATen/native/LinearAlgebra.cpp:1147]:
```suggestion
    return self * beta;
```

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 19]
```diff
@@ -1147,10 +1162,6 @@ static void addmm_impl_cpu_(
     return;
   }
 
-  if (beta.toComplexDouble() != 0.0 && !self.is_same(result)) {
```
Comment (anjali411) [inline aten/src/ATen/native/LinearAlgebra.cpp:1150]: please undo this change

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 17 above)
Comment (HaoZeke) [reply aten/src/ATen/native/LinearAlgebra.cpp:1141]: This is a `void` function

File: aten/src/ATen/native/cuda/Blas.cpp [hunk 20]
```diff
@@ -181,6 +181,19 @@ Tensor& addmm_out_cuda_impl(Tensor& result, const Tensor& self, const Tensor& ma
   }
 
   if (&result != &self) {
+
+    // beta*self + alpha*(mat1@mat2) = 0
+    if ((mat1._is_zerotensor() || mat2._is_zerotensor() || alpha.equal(0.0)) &&
+        (self._is_zerotensor() || beta.equal(0.0))) {
+      result = at::_efficientzerotensor(self_sizes, self.options());
```
Comment (anjali411) [inline aten/src/ATen/native/cuda/Blas.cpp:188]: pssht too bad we already allocated result before we reach here. it's ok though.

**Comment by HaoZeke** [comment]
Marking as draft until performance gains are clearer.

```
Using pytorch 1.12.0a0+git41b7460
50 / 50[------------------------ FWD mode AD input dtype:torch.float32 device:cpu ------------------------]
                    |  ZT->inp, mat1, mat2  |  ZT->mat1, mat2  |  ZT->mat2  |  ZT->inp   |   No ZT  
1 threads: -----------------------------------------------------------------------------------------
      (32, 32)      |            58.8       |         59.0     |      59.0  |      58.9  |      58.9
      (64, 64)      |            75.2       |         76.1     |      76.4  |      76.3  |      76.2
      (128, 128)    |           154.3       |        155.5     |     155.2  |     155.5  |     155.4
      (256, 256)    |           563.7       |        564.6     |     565.0  |     565.2  |     565.0
      (512, 512)    |          2656.7       |       2655.5     |    2660.5  |    2665.3  |    2661.1
      (1024, 1024)  |         15331.0       |      15298.8     |   15230.9  |   15338.6  |   15269.3
      (2048, 2048)  |        107012.0       |     106929.7     |  107141.0  |  106853.9  |  106587.9
      (4096, 4096)  |        742985.0       |     750298.1     |  749165.8  |  749519.6  |  745305.1

Times are in microseconds (us).

Using pytorch 1.12.0a0+git41b7460
50 / 50[-------------------------- FWD mode AD input dtype:torch.float64 device:cpu -------------------------]
                    |  ZT->inp, mat1, mat2  |  ZT->mat1, mat2  |   ZT->mat2  |   ZT->inp   |    No ZT  
1 threads: --------------------------------------------------------------------------------------------
      (32, 32)      |            65.2       |         65.1     |       65.2  |       65.2  |       65.0
      (64, 64)      |            97.5       |         98.6     |       98.7  |       98.6  |       98.6
      (128, 128)    |           268.3       |        273.4     |      270.4  |      274.5  |      287.0
      (256, 256)    |          1098.2       |       1101.5     |     1103.4  |     1099.6  |     1095.8
      (512, 512)    |          5245.7       |       5245.7     |     5240.4  |     5244.7  |     5246.1
      (1024, 1024)  |         31116.6       |      31118.4     |    31147.5  |    31129.4  |    31129.4
      (2048, 2048)  |        216907.6       |     216341.7     |   216934.8  |   216953.3  |   215885.5
      (4096, 4096)  |       1523638.4       |    1519501.2     |  1521577.6  |  1529437.7  |  1527426.4

Times are in microseconds (us).
```

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 21]
```diff
@@ -1134,6 +1134,22 @@ static void addmm_impl_cpu_(
     return;
   }
 
+  if (!result._is_zerotensor() and (m1._is_zerotensor() or m2._is_zerotensor())){
```
Comment (anjali411) [inline aten/src/ATen/native/LinearAlgebra.cpp:1137]: this is incorrect. you need a separate error check. please revert to the old change for the rest

`TORCH_CHECK(!!result._is_zerotensor(), ...);`
